[python3] `unordered_compare` does support comparing list with dictionaries

In python3 virtualenv, below comparsion is failed:
~~~
result = [{'nearfull_percent': '87%', 'osd_name': 'osd.1'}, {'nearfull_percent': '86%', 'osd_name': 'osd.2'}]
expected = [{'nearfull_percent': '87%', 'osd_name': 'osd.1'}, {'nearfull_percent': '86%', 'osd_name': 'osd.2'}]

    def unordered_compare(result, expected):
        """
        Deep compare rule reducer results when testing.  Developed to find
        arbitrarily nested lists and remove differences based on ordering.
        """
        logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)

        if isinstance(result, dict) and expected is None:
            assert result["type"] == "skip", result
            return

        if not all(isinstance(o, six.string_types) for o in (result, expected)):
            assert issubclass(type(result), type(expected))

        if isinstance(result, list):
            assert len(result) == len(expected)
>           for left_item, right_item in six.moves.zip(sorted(result), sorted(expected)):
E           TypeError: '<' not supported between instances of 'dict' and 'dict'
~~~

Tuple should be supported
~~~
d = ('default', 'policy-test', ['www.example.com'])

    def _sort_key(d):
        """
            Deep sort the elements which is in different types when comparing two
            lists.

            """
        dt = []
        if isinstance(d, dict):
            dt = list(d.items())
        elif isinstance(d, (set, tuple)):
>           dt = sorted(d)
E           TypeError: '<' not supported between instances of 'list' and 'str'
~~~

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -7,7 +7,7 @@ import logging
 import six
 import six.moves
 from collections import defaultdict
-from functools import wraps, cmp_to_key
+from functools import wraps
 
 try:
     from StringIO import StringIO
@@ -77,7 +77,7 @@ def unordered_compare(result, expected):
         dt = []
         if isinstance(d, dict):
             dt = list(d.items())
-        elif isinstance(d, (set, tuple, range)):
+        elif isinstance(d, (set, tuple)):
             dt = sorted(d)
         elif isinstance(d, list):
             dt = sorted(d, key=_sort_key)

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -7,7 +7,7 @@ import logging
 import six
 import six.moves
 from collections import defaultdict
-from functools import wraps
+from functools import wraps, cmp_to_key
 
 try:
     from StringIO import StringIO
@@ -68,9 +68,22 @@ def unordered_compare(result, expected):
     Deep compare rule reducer results when testing.  Developed to find
     arbitrarily nested lists and remove differences based on ordering.
     """
-    def sort_key(d):
-        """ To support there are dictionary elements when comparing two lists."""
-        return sorted(d.items()) if isinstance(d, dict) else d
+    def _sort_key(d):
+        """
+        Deep sort the elements which is in different types when comparing two
+        lists.
+
+        """
+        dt = []
+        if isinstance(d, dict):
+            dt = list(d.items())
+        elif isinstance(d, (set, tuple, range)):
+            dt = sorted(d)
+        elif isinstance(d, list):
+            dt = sorted(d, key=_sort_key)
+        if dt:
+            return '{}'.format(dt[0])
+        return '{}'.format(d)
 
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)
 
@@ -83,8 +96,8 @@ def unordered_compare(result, expected):
 
     if isinstance(result, list):
         assert len(result) == len(expected)
-        result = sorted(result, key=sort_key)
-        expected = sorted(expected, key=sort_key)
+        result = sorted(result, key=_sort_key)
+        expected = sorted(expected, key=_sort_key)
         for left_item, right_item in six.moves.zip(result, expected):
             unordered_compare(left_item, right_item)
     elif isinstance(result, dict):

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -70,16 +70,16 @@ def unordered_compare(result, expected):
     """
     def _sort_key(d):
         """
-        Deep sort the elements which is in different types when comparing two
+        Deep sort the elements which are in different types when comparing two
         lists.
 
         """
         dt = []
         if isinstance(d, dict):
             dt = list(d.items())
-        elif isinstance(d, (set, tuple)):
+        elif isinstance(d, set):
             dt = sorted(d)
-        elif isinstance(d, list):
+        elif isinstance(d, (list, tuple)):
             dt = sorted(d, key=_sort_key)
         if dt:
             return '{}'.format(dt[0])

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -68,6 +68,9 @@ def unordered_compare(result, expected):
     Deep compare rule reducer results when testing.  Developed to find
     arbitrarily nested lists and remove differences based on ordering.
     """
+    def sort_key(d):
+        return sorted(d.items()) if isinstance(d, dict) else d
+
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)
 
     if isinstance(result, dict) and expected is None:
@@ -80,9 +83,8 @@ def unordered_compare(result, expected):
     if isinstance(result, list):
         assert len(result) == len(expected)
         # To support there are elements in type of dictionary
-        sorted_key = lambda d: sorted(d.items()) if isinstance(d, dict) else d
-        result = sorted(result, key=sorted_key)
-        expected = sorted(expected,  key=sorted_key)
+        result = sorted(result, key=sort_key)
+        expected = sorted(expected, key=sort_key)
         for left_item, right_item in six.moves.zip(result, expected):
             unordered_compare(left_item, right_item)
     elif isinstance(result, dict):

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -8,6 +8,7 @@ import six
 import six.moves
 from collections import defaultdict
 from functools import wraps
+from operator import eq
 
 try:
     from StringIO import StringIO
@@ -73,20 +74,7 @@ def deep_compare(result, expected):
         assert result["type"] == "skip", result
         return
 
-    if not all(isinstance(o, six.string_types) for o in (result, expected)):
-        assert issubclass(type(result), type(expected))
-
-    if isinstance(result, (list, tuple)):
-        assert len(result) == len(expected)
-        for left_item, right_item in six.moves.zip(result, expected):
-            deep_compare(left_item, right_item)
-    elif isinstance(result, dict):
-        assert len(result) == len(expected)
-        for item_key in result:
-            assert item_key in expected
-            deep_compare(result[item_key], expected[item_key])
-    else:
-        assert result == expected
+    assert eq(result, expected)
 
 
 def run_input_data(component, input_data):

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -76,7 +76,7 @@ def unordered_compare(result, expected):
         """
         dt = []
         if isinstance(d, dict):
-            dt = list(d.items())
+            dt = sorted(d.items())
         elif isinstance(d, set):
             dt = sorted(d)
         elif isinstance(d, (list, tuple)):

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -82,8 +82,8 @@ def unordered_compare(result, expected):
         elif isinstance(d, (list, tuple)):
             dt = sorted(d, key=_sort_key)
         if dt:
-            return '{}'.format(dt[0])
-        return '{}'.format(d)
+            return '{0}'.format(dt[0])
+        return '{0}'.format(d)
 
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)
 

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -94,7 +94,7 @@ def unordered_compare(result, expected):
     if not all(isinstance(o, six.string_types) for o in (result, expected)):
         assert issubclass(type(result), type(expected))
 
-    if isinstance(result, list):
+    if isinstance(result, (list, tuple)):
         assert len(result) == len(expected)
         result = sorted(result, key=_sort_key)
         expected = sorted(expected, key=_sort_key)

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -79,7 +79,11 @@ def unordered_compare(result, expected):
 
     if isinstance(result, list):
         assert len(result) == len(expected)
-        for left_item, right_item in six.moves.zip(sorted(result), sorted(expected)):
+        # To support there are elements in type of dictionary
+        sorted_key = lambda d: sorted(d.items()) if isinstance(d, dict) else d
+        result = sorted(result, key=sorted_key)
+        expected = sorted(expected,  key=sorted_key)
+        for left_item, right_item in six.moves.zip(result, expected):
             unordered_compare(left_item, right_item)
     elif isinstance(result, dict):
         assert len(result) == len(expected)

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -69,6 +69,7 @@ def unordered_compare(result, expected):
     arbitrarily nested lists and remove differences based on ordering.
     """
     def sort_key(d):
+        """ To support there are elements in type of dictionary."""
         return sorted(d.items()) if isinstance(d, dict) else d
 
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)
@@ -82,7 +83,6 @@ def unordered_compare(result, expected):
 
     if isinstance(result, list):
         assert len(result) == len(expected)
-        # To support there are elements in type of dictionary
         result = sorted(result, key=sort_key)
         expected = sorted(expected, key=sort_key)
         for left_item, right_item in six.moves.zip(result, expected):

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -82,7 +82,7 @@ def unordered_compare(result, expected):
         elif isinstance(d, (list, tuple)):
             dt = sorted(d, key=_sort_key)
         if dt:
-            return '{0}'.format(dt[0])
+            return '{0}'.format(dt)
         return '{0}'.format(d)
 
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -63,28 +63,10 @@ DEFAULT_RELEASE = "Red Hat Enterprise Linux Server release 7.2 (Maipo)"
 DEFAULT_HOSTNAME = "hostname.example.com"
 
 
-def unordered_compare(result, expected):
+def deep_compare(result, expected):
     """
-    Deep compare rule reducer results when testing.  Developed to find
-    arbitrarily nested lists and remove differences based on ordering.
+    Deep compare rule reducer results when testing.
     """
-    def _sort_key(d):
-        """
-        Deep sort the elements which are in different types when comparing two
-        lists.
-
-        """
-        dt = []
-        if isinstance(d, dict):
-            dt = sorted(d.items())
-        elif isinstance(d, set):
-            dt = sorted(d)
-        elif isinstance(d, (list, tuple)):
-            dt = sorted(d, key=_sort_key)
-        if dt:
-            return '{0}'.format(dt)
-        return '{0}'.format(d)
-
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)
 
     if isinstance(result, dict) and expected is None:
@@ -96,14 +78,13 @@ def unordered_compare(result, expected):
 
     if isinstance(result, (list, tuple)):
         assert len(result) == len(expected)
-        result = sorted(result, key=_sort_key)
-        expected = sorted(expected, key=_sort_key)
         for left_item, right_item in six.moves.zip(result, expected):
-            unordered_compare(left_item, right_item)
+            deep_compare(left_item, right_item)
     elif isinstance(result, dict):
         assert len(result) == len(expected)
         for item_key in result:
-            unordered_compare(result[item_key], expected[item_key])
+            assert item_key in expected
+            deep_compare(result[item_key], expected[item_key])
     else:
         assert result == expected
 
@@ -134,7 +115,7 @@ def run_test(component, input_data, expected=None):
 
     broker = run_input_data(component, input_data)
     if expected:
-        unordered_compare(broker.get(component), expected)
+        deep_compare(broker.get(component), expected)
     return broker.get(component)
 
 
@@ -326,7 +307,7 @@ def redhat_release(major, minor=""):
         raise Exception("invalid major version: %s" % major)
 
 
-def archive_provider(component, test_func=unordered_compare, stride=1):
+def archive_provider(component, test_func=deep_compare, stride=1):
     """
     Decorator used to register generator functions that yield InputData and
     expected response tuples.  These generators will be consumed by py.test
@@ -347,7 +328,7 @@ def archive_provider(component, test_func=unordered_compare, stride=1):
         yield every `stride` InputData object rather than the full set. This
         is used to provide a faster execution path in some test setups.
 
-    [1] insights.tests.unordered_compare()
+    [1] insights.tests.deep_compare()
     """
     def _wrap(func):
         @six.wraps(func)

--- a/insights/tests/__init__.py
+++ b/insights/tests/__init__.py
@@ -69,7 +69,7 @@ def unordered_compare(result, expected):
     arbitrarily nested lists and remove differences based on ordering.
     """
     def sort_key(d):
-        """ To support there are elements in type of dictionary."""
+        """ To support there are dictionary elements when comparing two lists."""
         return sorted(d.items()) if isinstance(d, dict) else d
 
     logger.debug("--Comparing-- (%s) %s to (%s) %s", type(result), result, type(expected), expected)

--- a/insights/tests/test_test.py
+++ b/insights/tests/test_test.py
@@ -1,9 +1,9 @@
-from insights.tests import unordered_compare
+from insights.tests import deep_compare
 
 
-def test_unordered_compare():
+def test_deep_compare():
     """
-    Tests that unordered_compare() doesn't modify the objects it compares.
+    Tests that deep_compare() doesn't modify the objects it compares.
     """
     x_list = [8, 5, 3, 1, 2, 9, 5]
     x_list_backup = [8, 5, 3, 1, 2, 9, 5]
@@ -27,11 +27,11 @@ def test_unordered_compare():
     assert b_tuple == b_tuple_backup
     noexception = False
     try:
-        unordered_compare(a, b)
+        deep_compare(a, b)
         noexception = True
     except:
         pass
-    # unordered_compare test - the original objects should stay unchanged
+    # deep_compare test - the original objects should stay unchanged
     assert noexception
     assert a == b
     assert a == c

--- a/insights/tests/test_util.py
+++ b/insights/tests/test_util.py
@@ -102,6 +102,44 @@ def test_dict():
         unordered_compare({"foo": 1, "bar": [1, 2, 3]}, {"foo": 1, "bar": [0, 1, 2]})
 
 
+def test_nest_list_dict():
+    unordered_compare([{2: 22}, {3: 33}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
+    with pytest.raises(AssertionError):
+        unordered_compare([{2: 22}, {3: 30}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
+
+
+def test_nest_list_list():
+    unordered_compare([[2, 22, 3], [33, 4, 44]], [[33, 4, 44], [2, 22, 3]])
+    with pytest.raises(AssertionError):
+        unordered_compare([[2, 22], [33, 44, 4]], [[2, 22, 3], [33, 4, 44]])
+        unordered_compare([[2, 22, [3]], [33, 44, 4]], [[2, 22, 3], [33, 4, 44]])
+    unordered_compare([[2, [22, 3]], [33, 4, 44]], [[33, 4, 44], [2, [22, 3]]])
+    unordered_compare([[2, 22, 3], [], [33, 4, 44], []], [[], [], [33, 4, 44], [2, 22, 3]])
+    unordered_compare([[2, [22], [3, 33]], [], [33, 4, 44], []], [[], [], [33, 4, 44], [[22], 2, [3, 33]]])
+
+
+def test_nest_list_list_str():
+    unordered_compare([[2, 22, 3], 33, 4, 44], [33, 4, 44, [2, 22, 3]])
+    with pytest.raises(AssertionError):
+        unordered_compare([[2, 22], 33, 44, 4], [[2, 22, 3], 33, 4, 44])
+
+
+def test_nest_list_dict_str():
+    unordered_compare([{2: 22}, {3: 33}, 4, 44], [4, 44, {3: 33}, {2: 22}])
+    with pytest.raises(AssertionError):
+        unordered_compare([{2: 20}, {3: 33}, 4, 44], [4, 44, {3: 33}, {2: 22}])
+
+
+def test_nest_dict_list():
+    unordered_compare({2: [22, 222], 3: [33, 333]}, {2: [222, 22], 3: [33, 333]})
+    with pytest.raises(AssertionError):
+        unordered_compare({2: [22, 22], 3: [33, 333]}, {2: [22, 222], 3: [33, 333]})
+    unordered_compare({2: (22, 222), 3: [33, 333]}, {2: (22, 222), 3: [333, 33]})
+    unordered_compare({3: [33, 333], 2: (22, 222)}, {2: (22, 222), 3: [333, 33]})
+    unordered_compare({3: [33, [3, '33', 333]], 2: (22, 222)}, {2: (22, 222), 3: [[3, '33', 333], 33]})
+    unordered_compare({3: [33, [set([4, 3]), '33', 333]], 2: (22, 222)}, {2: (22, 222), 3: [['33', 333, set([3, 4])], 33]})
+
+
 def test_deep_nest():
     a = {"error_key": "test1", "stuff": {"abba": [{"foo": 2}]}}
     b = {"error_key": "test1", "stuff": {"abba": [{"foo": 2}]}}

--- a/insights/tests/test_util.py
+++ b/insights/tests/test_util.py
@@ -1,6 +1,6 @@
 import pytest
 import warnings
-from insights.tests import unordered_compare
+from insights.tests import deep_compare
 from insights.core.dr import split_requirements, stringify_requirements, get_missing_requirements
 from insights.core import context
 from insights.util import case_variants, deprecated
@@ -72,107 +72,126 @@ def test_non_product_member(ctx):
 
 
 def test_str():
-    unordered_compare("foo", "foo")
-    unordered_compare(u"foo", u"foo")
+    deep_compare("foo", "foo")
+    deep_compare(u"foo", u"foo")
     with pytest.raises(AssertionError):
-        unordered_compare("foo", "bar")
+        deep_compare("foo", "bar")
 
 
 def test_num():
-    unordered_compare(1, 1)
-    unordered_compare(1.1, 1.1)
+    deep_compare(1, 1)
+    deep_compare(1.1, 1.1)
     with pytest.raises(AssertionError):
-        unordered_compare(1, 2)
+        deep_compare(1, 2)
 
 
 def test_list():
-    unordered_compare([1, 2, 3], [1, 2, 3])
-    unordered_compare([2, 3, 4], [4, 3, 2])
+    deep_compare([1, 2, 3], [1, 2, 3])
     with pytest.raises(AssertionError):
-        unordered_compare([1, 2, 3], [2, 3, 4])
+        deep_compare([2, 3, 4], [4, 3, 2])
+        deep_compare([1, 2, 3], [2, 3, 4])
 
 
 def test_dict():
-    unordered_compare({"foo": 1}, {"foo": 1})
-    unordered_compare({"foo": 1, "bar": 2}, {"bar": 2, "foo": 1})
+    deep_compare({"foo": 1}, {"foo": 1})
+    deep_compare({"foo": 1, "bar": 2}, {"bar": 2, "foo": 1})
     with pytest.raises(AssertionError):
-        unordered_compare({"foo": 1}, {"foo": 2})
+        deep_compare({"foo": 1}, {"foo": 2})
 
     with pytest.raises(AssertionError):
-        unordered_compare({"foo": 1, "bar": [1, 2, 3]}, {"foo": 1, "bar": [0, 1, 2]})
-
-
-def test_nest_tuple():
-    unordered_compare(({2: 22}, {3: 33}, {4: 44}), ({2: 22}, {3: 33}, {4: 44}))
-    with pytest.raises(AssertionError):
-        unordered_compare(({2: 22}, {3: 30}, {4: 44}), ({4: 44}, {3: 33}, {2: 22}))
-    unordered_compare(({2: 22}, {3: [33, 333]}, {4: 44}), ({4: 44}, {3: [333, 33]}, {2: 22}))
-    unordered_compare(({2: 22}, [33, 333, '3'], {4: 44}), ([333, '3', 33], {4: 44}, {2: 22}))
-
-
-def test_nest_list_dict():
-    unordered_compare([{2: 22}, {3: 33}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
-    with pytest.raises(AssertionError):
-        unordered_compare([{2: 22}, {3: 30}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
-    unordered_compare([{2: 22}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {2: 22}])
-    unordered_compare([{2: 22, 6: 66}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {6: 66, 2: 22}])
-    unordered_compare(
-            [
-                {'ip_addr': '0.0.0.0', 'process_name': 'qpidd', 'port': '5672'},
-                {'ip_addr': '127.0.0.1', 'process_name': 'mongod', 'port': '27017'},
-                {'ip_addr': '127.0.0.1', 'process_name': 'Passenger Rac', 'port': '53644'},
-                {'ip_addr': '0.0.0.0', 'process_name': 'qdrouterd', 'port': '5646'}
-            ],
-            [
-                {'ip_addr': '0.0.0.0', 'port': '5646', 'process_name': 'qdrouterd'},
-                {'ip_addr': '0.0.0.0', 'port': '5672', 'process_name': 'qpidd'},
-                {'ip_addr': '127.0.0.1', 'port': '53644', 'process_name': 'Passenger Rac'},
-                {'ip_addr': '127.0.0.1', 'port': '27017', 'process_name': 'mongod'}
-            ]
-    )
-
-
-def test_nest_list_list():
-    unordered_compare([[2, 22, 3], [33, 4, 44]], [[33, 4, 44], [2, 22, 3]])
-    with pytest.raises(AssertionError):
-        unordered_compare([[2, 22], [33, 44, 4]], [[2, 22, 3], [33, 4, 44]])
-        unordered_compare([[2, 22, [3]], [33, 44, 4]], [[2, 22, 3], [33, 4, 44]])
-    unordered_compare([[2, [22, 3]], [33, 4, 44]], [[33, 4, 44], [2, [22, 3]]])
-    unordered_compare([[2, 22, 3], [], [33, 4, 44], []], [[], [], [33, 4, 44], [2, 22, 3]])
-    unordered_compare([[2, [22], [3, 33]], [], [33, 4, 44], []], [[], [], [33, 4, 44], [[22], 2, [3, 33]]])
-
-
-def test_nest_list_list_str():
-    unordered_compare([[2, 22, 3], 33, 4, 44], [33, 4, 44, [2, 22, 3]])
-    with pytest.raises(AssertionError):
-        unordered_compare([[2, 22], 33, 44, 4], [[2, 22, 3], 33, 4, 44])
-
-
-def test_nest_list_dict_str():
-    unordered_compare([{2: 22}, {3: 33}, 4, 44], [4, 44, {3: 33}, {2: 22}])
-    with pytest.raises(AssertionError):
-        unordered_compare([{2: 20}, {3: 33}, 4, 44], [4, 44, {3: 33}, {2: 22}])
-
-
-def test_nest_dict_list():
-    unordered_compare({2: [22, 222], 3: [33, 333]}, {2: [222, 22], 3: [33, 333]})
-    with pytest.raises(AssertionError):
-        unordered_compare({2: [22, 22], 3: [33, 333]}, {2: [22, 222], 3: [33, 333]})
-    unordered_compare({2: (22, 222), 3: [33, 333]}, {2: (22, 222), 3: [333, 33]})
-    unordered_compare({3: [33, 333], 2: (22, 222)}, {2: (22, 222), 3: [333, 33]})
-    unordered_compare({3: [33, [3, '33', 333]], 2: (22, 222)}, {2: (22, 222), 3: [[3, '33', 333], 33]})
-    unordered_compare({3: [33, [set([4, 3]), '33', 333]], 2: (22, 222)}, {2: (22, 222), 3: [['33', 333, set([3, 4])], 33]})
+        deep_compare({"foo": 1, "bar": [1, 2, 3]}, {"foo": 1, "bar": [0, 1, 2]})
 
 
 def test_deep_nest():
     a = {"error_key": "test1", "stuff": {"abba": [{"foo": 2}]}}
     b = {"error_key": "test1", "stuff": {"abba": [{"foo": 2}]}}
 
-    unordered_compare(a, b)
+    deep_compare(a, b)
 
     with pytest.raises(AssertionError):
         b["stuff"]["abba"][0]["foo"] = "cake"
-        unordered_compare(a, b)
+        deep_compare(a, b)
+
+
+def test_deep_nest_tuple_dict():
+    a = ({2: 22}, {3: 33}, {4: 44})
+    b = ({2: 22}, {3: 33}, {4: 44})
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b = ({2: 22}, {3: 30}, {4: 44})
+        deep_compare(a, b)
+
+    a = ({2: 22}, {3: [33, 333]}, {4: 44})
+    b = ({2: 22}, {3: [33, 333]}, {4: 44})
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b = ({2: 22}, {3: [333, 33]}, {4: 44})
+        deep_compare(a, b)
+
+    a = ({2: 22, 5: '55'}, [33, 333, '3'], {6: [6], 4: 44})
+    b = ({5: '55', 2: 22}, [33, 333, '3'], {4: 44, 6: [6]})
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b = ({5: '55', 2: 22}, [33, 333, 3], {4: 44, 6: [6]})
+        deep_compare(a, b)
+
+
+def test_deep_nest_list_dict():
+    a = [{2: 22}, {3: 33}, {4: 44}]
+    b = [{2: 22}, {3: 33}, {4: 44}]
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b[1][3] = 30
+        deep_compare(a, b)
+
+    a = [{2: 22}, {3: [33, 333]}, {4: 44}]
+    b = [{2: 22}, {3: [33, 333]}, {4: 44}]
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b[1][3] = [333, 33]
+        deep_compare(a, b)
+
+    a = [{2: 22, 5: '55'}, [33, 333, '3'], {6: [6], 4: 44}]
+    b = [{5: '55', 2: 22}, [33, 333, '3'], {4: 44, 6: [6]}]
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b[1][2] = 3
+        deep_compare(a, b)
+
+    a = [
+           {'ip_addr': '0.0.0.0', 'process_name': 'qpidd', 'port': '5672'},
+           {'ip_addr': '127.0.0.1', 'process_name': 'mongod', 'port': '27017'},
+           {'ip_addr': '127.0.0.1', 'process_name': 'Passenger Rac', 'port': '53644'},
+           {'ip_addr': '0.0.0.0', 'process_name': 'qdrouterd', 'port': '5646'},
+    ]
+    b = [
+           {'ip_addr': '0.0.0.0', 'port': '5672', 'process_name': 'qpidd'},
+           {'ip_addr': '127.0.0.1', 'port': '27017', 'process_name': 'mongod'},
+           {'ip_addr': '127.0.0.1', 'port': '53644', 'process_name': 'Passenger Rac'},
+           {'ip_addr': '0.0.0.0', 'port': '5646', 'process_name': 'qdrouterd'},
+    ]
+    deep_compare(a, b)
+
+
+def test_deep_nest_dict_list():
+    a = {'2': [22, 222], 3: [33, 333]}
+    b = {3: [33, 333], '2': [22, 222]}
+    deep_compare(a, b)
+    with pytest.raises(AssertionError):
+        b['2'] = [222, 22]
+        deep_compare(a, b)
+
+    a = {3: [33, [3, '33', 333]], 2: (22, 222)}
+    b = {2: (22, 222), 3: [33, [3, '33', 333]]}
+    with pytest.raises(AssertionError):
+        b[3] = [33, ['33', 333, 3]]
+        deep_compare(a, b)
+
+    a = {3: [33, [set([4, 3]), '33', 333]], 2: (22, 222)}
+    b = {2: (22, 222), 3: [33, [set([4, 3]), '33', 333]]}
+    with pytest.raises(AssertionError):
+        b[3][1] = set([3, 4])
+        deep_compare(a, b)
 
 
 def test_case_variants():

--- a/insights/tests/test_util.py
+++ b/insights/tests/test_util.py
@@ -116,6 +116,20 @@ def test_nest_list_dict():
         unordered_compare([{2: 22}, {3: 30}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
     unordered_compare([{2: 22}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {2: 22}])
     unordered_compare([{2: 22, 6: 66}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {6: 66, 2: 22}])
+    unordered_compare(
+            [
+                {'ip_addr': '0.0.0.0', 'process_name': 'qpidd', 'port': '5672'},
+                {'ip_addr': '127.0.0.1', 'process_name': 'mongod', 'port': '27017'},
+                {'ip_addr': '127.0.0.1', 'process_name': 'Passenger Rac', 'port': '53644'},
+                {'ip_addr': '0.0.0.0', 'process_name': 'qdrouterd', 'port': '5646'}
+            ],
+            [
+                {'ip_addr': '0.0.0.0', 'port': '5646', 'process_name': 'qdrouterd'},
+                {'ip_addr': '0.0.0.0', 'port': '5672', 'process_name': 'qpidd'},
+                {'ip_addr': '127.0.0.1', 'port': '53644', 'process_name': 'Passenger Rac'},
+                {'ip_addr': '127.0.0.1', 'port': '27017', 'process_name': 'mongod'}
+            ]
+    )
 
 
 def test_nest_list_list():

--- a/insights/tests/test_util.py
+++ b/insights/tests/test_util.py
@@ -115,6 +115,7 @@ def test_nest_list_dict():
     with pytest.raises(AssertionError):
         unordered_compare([{2: 22}, {3: 30}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
     unordered_compare([{2: 22}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {2: 22}])
+    unordered_compare([{2: 22, 6: 66}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {6: 66, 2: 22}])
 
 
 def test_nest_list_list():

--- a/insights/tests/test_util.py
+++ b/insights/tests/test_util.py
@@ -109,6 +109,7 @@ def test_nest_tuple():
     unordered_compare(({2: 22}, {3: [33, 333]}, {4: 44}), ({4: 44}, {3: [333, 33]}, {2: 22}))
     unordered_compare(({2: 22}, [33, 333, '3'], {4: 44}), ([333, '3', 33], {4: 44}, {2: 22}))
 
+
 def test_nest_list_dict():
     unordered_compare([{2: 22}, {3: 33}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
     with pytest.raises(AssertionError):

--- a/insights/tests/test_util.py
+++ b/insights/tests/test_util.py
@@ -102,10 +102,18 @@ def test_dict():
         unordered_compare({"foo": 1, "bar": [1, 2, 3]}, {"foo": 1, "bar": [0, 1, 2]})
 
 
+def test_nest_tuple():
+    unordered_compare(({2: 22}, {3: 33}, {4: 44}), ({2: 22}, {3: 33}, {4: 44}))
+    with pytest.raises(AssertionError):
+        unordered_compare(({2: 22}, {3: 30}, {4: 44}), ({4: 44}, {3: 33}, {2: 22}))
+    unordered_compare(({2: 22}, {3: [33, 333]}, {4: 44}), ({4: 44}, {3: [333, 33]}, {2: 22}))
+    unordered_compare(({2: 22}, [33, 333, '3'], {4: 44}), ([333, '3', 33], {4: 44}, {2: 22}))
+
 def test_nest_list_dict():
     unordered_compare([{2: 22}, {3: 33}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
     with pytest.raises(AssertionError):
         unordered_compare([{2: 22}, {3: 30}, {4: 44}], [{4: 44}, {3: 33}, {2: 22}])
+    unordered_compare([{2: 22}, {3: [33, 333]}, {4: 44}], [{4: 44}, {3: [333, 33]}, {2: 22}])
 
 
 def test_nest_list_list():

